File: reorder_list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *next = nullptr, *prev = nullptr;
        
        while(head) {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
    void reorderList(ListNode* head) {
        ListNode *slow = head, *fast = head;

        while(fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* back = reverseList(slow);
        ListNode* front = head;

        while(front->next && back->next) {
            ListNode *temp1 = front, *temp2 = back;

            front = front->next;
            back = back->next;
            temp1->next = temp2;
            temp2->next = front;
        }
    }
};

// Time Complexity : O(n)
// Space Complexity : O(1)

File: reorder_list.go
func reverseList(head *ListNode) *ListNode {
    var next, prev *ListNode = nil, nil
    
    for head != nil {
        next = head.Next
        head.Next = prev
        prev = head
        head = next
    }
    return prev
}

func reorderList(head *ListNode) {
    slow, fast := head, head

    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    back := reverseList(slow)
    front := head

    for front.Next != nil && back.Next != nil {
        temp1, temp2 := front, back

        front = front.Next
        back = back.Next
        temp1.Next = temp2
        temp2.Next = front
    }
}

// Time Complexity : O(n)
// Space Complexity : O(1)